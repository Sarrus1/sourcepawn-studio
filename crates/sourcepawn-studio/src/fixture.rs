use anyhow::Result;
use crossbeam::channel::Receiver;
use itertools::Itertools;
use lsp_server::{Connection, Response};
use lsp_types::{
    notification::{DidOpenTextDocument, Exit, Initialized},
    request::{Completion, Initialize, ResolveCompletionItem, Shutdown},
    ClientCapabilities, CompletionContext, CompletionItem, CompletionItemKind, CompletionParams,
    CompletionResponse, CompletionTriggerKind, DidOpenTextDocumentParams, InitializeParams,
    InitializedParams, Location, LocationLink, Position, Range, SignatureHelp,
    TextDocumentIdentifier, TextDocumentItem, TextDocumentPositionParams, Url, WorkspaceFolder,
};
use std::{
    env,
    fs::File,
    io,
    path::{Path, PathBuf},
    str::FromStr,
    sync::Once,
    thread::JoinHandle,
    time::Duration,
};
use tempfile::{tempdir, TempDir};
use zip::ZipArchive;

use crate::{config::ConfigData, lsp};

use super::{GlobalState, LspClient};

#[derive(Debug)]
pub enum InternalMessage {
    OptionsRequested,
}

#[derive(Debug)]
pub struct Fixture {
    pub documents: Vec<Document>,
}

impl Fixture {
    pub fn parse(input: &str) -> Fixture {
        let mut documents = Vec::new();

        let mut start = 0;
        if !input.is_empty() {
            for end in input
                .match_indices("%!")
                .skip(1)
                .map(|(i, _)| i)
                .chain(std::iter::once(input.len()))
            {
                documents.push(Document::parse(&input[start..end]));
                start = end;
            }
        }

        Self { documents }
    }

    pub fn write_files(&self, dir: &Path) {
        for document in &self.documents {
            let text = String::from(&document.text);
            let path = dir.join(&document.path);
            std::fs::create_dir_all(path.parent().unwrap()).unwrap();
            std::fs::write(&path, &text).unwrap();
        }
    }

    pub fn setup(&self, client: &LspClient, dir: &Path) {
        for document in &self.documents {
            let text = String::from(&document.text);
            let path = dir.join(&document.path);

            let uri = Url::from_file_path(&path).unwrap();
            let language_id = "sourcepawn".to_string();

            client
                .send_notification::<DidOpenTextDocument>(DidOpenTextDocumentParams {
                    text_document: TextDocumentItem::new(uri, language_id, 0, text),
                })
                .unwrap();
        }
    }
}

#[derive(Debug)]
pub struct Document {
    pub path: PathBuf,
    pub text: String,
    pub cursor: Option<Position>,
    pub ranges: Vec<Range>,
}

impl Document {
    pub fn parse(input: &str) -> Self {
        let mut lines = Vec::new();

        let (path, input) = input
            .trim()
            .strip_prefix("%! ")
            .map(|input| input.split_once('\n').unwrap_or((input, "")))
            .unwrap();

        let mut ranges = Vec::new();
        let mut cursor = None;

        for line in input.lines() {
            if line.chars().all(|c| matches!(c, ' ' | '^' | '|' | '!')) && !line.is_empty() {
                let index = (lines.len() - 1) as u32;

                cursor = cursor.or_else(|| {
                    let character = line.find('|')?;
                    Some(Position::new(index, character as u32))
                });

                if let Some(start) = line.find('!') {
                    let position = Position::new(index, start as u32);
                    ranges.push(Range::new(position, position));
                }

                if let Some(start) = line.find('^') {
                    let end = line.rfind('^').unwrap() + 1;
                    ranges.push(Range::new(
                        Position::new(index, start as u32),
                        Position::new(index, end as u32),
                    ));
                }
            } else {
                lines.push(line);
            }
        }

        Self {
            path: PathBuf::from(path),
            text: lines.join("\n"),
            cursor,
            ranges,
        }
    }
}

static LOGGER: Once = Once::new();

#[derive(Debug)]
pub struct TestBed {
    fixture: Fixture,
    locations: Vec<Location>,
    _temp_dir: TempDir,
    temp_dir_path: PathBuf,
    #[allow(unused)]
    temp_sm_dir_path: PathBuf,
    pub internal_rx: Receiver<InternalMessage>,
    client: LspClient,
    client_thread: Option<JoinHandle<()>>,
    server_thread: Option<JoinHandle<()>>,
}

impl Drop for TestBed {
    fn drop(&mut self) {
        log::trace!("Dropping server.");
        let _ = self.client.send_request::<Shutdown>(());
        let _ = self.client.send_notification::<Exit>(());
        self.client_thread.take().unwrap().join().unwrap();
        self.server_thread.take().unwrap().join().unwrap();
    }
}

impl TestBed {
    pub fn new(fixture: &str, add_sourcemod: bool) -> Result<Self> {
        LOGGER.call_once(|| {
            if option_env!("TEST_LOG") == Some("1") {
                fern::Dispatch::new()
                    .level(log::LevelFilter::Trace)
                    .chain(std::io::stderr())
                    .apply()
                    .unwrap()
            }
        });

        let fixture = Fixture::parse(fixture);

        let temp_dir = tempdir()?;
        let temp_dir_path = dunce::canonicalize(temp_dir.path())?;

        let temp_sm_dir = tempdir()?;
        let temp_sm_dir_path = dunce::canonicalize(temp_sm_dir.path())?;
        let temp_sm_dir_path_ = temp_sm_dir_path.clone(); // Copy the value to be able to move it into the closure

        let locations: Vec<Location> = fixture
            .documents
            .iter()
            .flat_map(|document| {
                let uri = Url::from_file_path(temp_dir_path.join(&document.path)).unwrap();
                document
                    .ranges
                    .iter()
                    .map(move |range| Location::new(uri.clone(), *range))
            })
            .collect();

        let (server_conn, client_conn) = Connection::memory();
        let (internal_tx, internal_rx) = crossbeam::channel::unbounded();

        let client = LspClient::new(client_conn.sender);

        let server_thread =
            std::thread::spawn(move || GlobalState::new(server_conn, false).run().unwrap());
        let client_thread = {
            let client = client.clone();
            std::thread::spawn(move || {
                let destination = temp_sm_dir_path_;
                for message in &client_conn.receiver {
                    match message {
                        lsp_server::Message::Request(request) => {
                            if request.method == "workspace/configuration" {
                                #[allow(unused_mut)]
                                let mut config = ConfigData::default();
                                if add_sourcemod {
                                    let mut current_dir = env::current_dir().unwrap();
                                    // The env depends on if we use the debugger or not.
                                    if current_dir.ends_with("sourcepawn-vscode") {
                                        current_dir = current_dir.join("crates/sourcepawn-studio/");
                                    }
                                    let sourcemod_path =
                                        current_dir.join("test_data/sourcemod.zip");
                                    unzip_file(&sourcemod_path, &destination).unwrap();
                                    #[cfg(test)]
                                    config
                                        .includeDirectories
                                        .push(destination.clone().join("include/"));
                                }
                                client
                                    .send_response(Response::new_ok(request.id, vec![config]))
                                    .unwrap();
                                std::thread::sleep(Duration::from_millis(100));
                                internal_tx.send(InternalMessage::OptionsRequested).unwrap()
                            } else {
                                client
                                    .send_error(
                                        request.id,
                                        lsp_server::ErrorCode::MethodNotFound,
                                        "Method not found".into(),
                                    )
                                    .unwrap();
                            }
                        }
                        lsp_server::Message::Response(response) => {
                            client.recv_response(response).unwrap();
                        }
                        lsp_server::Message::Notification(_) => {}
                    }
                }
            })
        };
        fixture.write_files(&temp_dir_path);

        Ok(TestBed {
            fixture,
            locations,
            _temp_dir: temp_dir,
            temp_dir_path,
            temp_sm_dir_path,
            client,
            internal_rx,
            client_thread: Some(client_thread),
            server_thread: Some(server_thread),
        })
    }

    pub fn initialize(&self, capabilities: ClientCapabilities) -> Result<()> {
        self.client
            .send_request::<Initialize>(InitializeParams {
                capabilities,
                workspace_folders: Some(vec![WorkspaceFolder {
                    uri: Url::from_file_path(self.temp_dir_path.clone()).unwrap(),
                    name: "test".to_string(),
                }]),
                root_uri: Some(Url::from_file_path(self.temp_dir_path.clone()).unwrap()),
                ..Default::default()
            })
            .unwrap();
        self.client
            .send_notification::<Initialized>(InitializedParams {})
            .unwrap();

        self.fixture.setup(&self.client, self.directory());

        Ok(())
    }

    pub fn client(&self) -> &LspClient {
        &self.client
    }

    pub fn cursor(&self) -> Option<TextDocumentPositionParams> {
        let (document, cursor) = self
            .fixture
            .documents
            .iter()
            .find_map(|document| document.cursor.map(|cursor| (document, cursor)))?;

        let uri = Url::from_file_path(self.temp_dir_path.join(&document.path)).unwrap();
        let id = TextDocumentIdentifier::new(uri);
        Some(TextDocumentPositionParams::new(id, cursor))
    }

    #[allow(unused)]
    pub fn locations(&self) -> &[Location] {
        &self.locations
    }

    pub fn directory(&self) -> &Path {
        &self.temp_dir_path
    }

    #[allow(unused)]
    pub fn documents(&self) -> &[Document] {
        &self.fixture.documents
    }

    /// Remove the tempdir path from the uri, so that the tests are not dependent
    /// on the tempdir.
    pub fn anonymize_uri(&self, uri: &mut Url) {
        let mut target_path = uri.to_file_path().unwrap();
        target_path = target_path
            .strip_prefix(&self.temp_dir_path)
            .unwrap()
            .to_path_buf();
        uri.set_path(&target_path.to_string_lossy());
    }
}

pub fn goto_definition(fixture: &str) -> Vec<LocationLink> {
    let test_bed = TestBed::new(fixture, true).unwrap();
    test_bed
        .initialize(
            serde_json::from_value(serde_json::json!({
                "textDocument": {
                    "definition": {
                        "linkSupport": true
                    }
                },
                "workspace": {
                    "configuration": true,
                    "workspace_folders": true
                }
            }))
            .unwrap(),
        )
        .unwrap();
    let text_document_position = test_bed.cursor().unwrap();
    let params = lsp_types::request::GotoTypeDefinitionParams {
        text_document_position_params: text_document_position,
        work_done_progress_params: Default::default(),
        partial_result_params: Default::default(),
    };
    let mut locations = match test_bed
        .client()
        .send_request::<lsp_types::request::GotoDefinition>(params)
        .unwrap()
    {
        Some(lsp_types::GotoDefinitionResponse::Link(locations)) => locations,
        _ => unreachable!("Expected a link response."),
    };

    locations.iter_mut().for_each(|location| {
        test_bed.anonymize_uri(&mut location.target_uri);
    });

    locations
}

pub fn complete(fixture: &str, trigger_character: Option<String>) -> Vec<CompletionItem> {
    let test_bed = TestBed::new(fixture, true).unwrap();
    test_bed
        .initialize(
            serde_json::from_value(serde_json::json!({
                "textDocument": {
                    "completion": {
                        "completionItem": {
                            "documentationFormat": ["plaintext", "markdown"]
                        }
                    }
                },
                "workspace": {
                    "configuration": true,
                    "workspace_folders": true
                }
            }))
            .unwrap(),
        )
        .unwrap();
    let text_document_position = test_bed.cursor().unwrap();
    let mut items = match test_bed
        .client()
        .send_request::<Completion>(CompletionParams {
            text_document_position,
            partial_result_params: Default::default(),
            work_done_progress_params: Default::default(),
            context: Some(CompletionContext {
                trigger_kind: CompletionTriggerKind::TRIGGER_CHARACTER,
                trigger_character,
            }),
        })
        .unwrap()
    {
        Some(CompletionResponse::Array(items)) => items,
        Some(CompletionResponse::List(list)) => list.items,
        None => Vec::new(),
    };

    items = items
        .into_iter()
        .map(|item| match item.data {
            Some(_) => {
                let mut item = test_bed
                    .client()
                    .send_request::<ResolveCompletionItem>(item)
                    .unwrap();

                item.data = None;
                item
            }
            None => item,
        })
        .sorted_by(|item1, item2| item1.label.cmp(&item2.label))
        .collect();

    // The results of include completions will always change because the paths of tempdir changes.
    for item in &mut items {
        if matches!(
            item.kind,
            Some(CompletionItemKind::FILE) | Some(CompletionItemKind::FOLDER)
        ) {
            item.detail = item.detail.as_mut().map(|detail| {
                // Account for windows paths
                detail.replace('\\', "/")
            });
        }
    }

    items
}

pub fn signature_help(fixture: &str) -> SignatureHelp {
    let test_bed = TestBed::new(fixture, true).unwrap();
    test_bed
        .initialize(
            serde_json::from_value(serde_json::json!({
                "textDocument": {
                    "signatureHelpProvider": {}
                },
                "workspace": {
                    "configuration": true,
                    "workspace_folders": true
                }
            }))
            .unwrap(),
        )
        .unwrap();
    let text_document_position = test_bed.cursor().unwrap();
    let params = lsp_types::SignatureHelpParams {
        text_document_position_params: text_document_position,
        work_done_progress_params: Default::default(),
        context: None,
    };

    test_bed
        .client()
        .send_request::<lsp_types::request::SignatureHelpRequest>(params)
        .unwrap()
        .unwrap()
}

pub fn unzip_file(zip_file_path: &Path, destination: &Path) -> Result<(), io::Error> {
    let file = File::open(zip_file_path)?;
    let mut archive = ZipArchive::new(file)?;

    for i in 0..archive.len() {
        let mut file = archive.by_index(i)?;
        let dest_path = destination.join(file.name());

        if file.is_dir() {
            std::fs::create_dir_all(&dest_path)?;
        } else {
            if let Some(parent_dir) = std::path::Path::new(&dest_path).parent() {
                std::fs::create_dir_all(parent_dir)?;
            }
            let mut dest_file = File::create(&dest_path)?;
            io::copy(&mut file, &mut dest_file)?;
        }
    }

    Ok(())
}

pub fn hover(fixture: &str) -> lsp::ext::Hover {
    let test_bed = TestBed::new(fixture, true).unwrap();
    test_bed
        .initialize(
            serde_json::from_value(serde_json::json!({
                "textDocument": {
                    "hover": {
                        "contentFormat": [
                            "plaintext",
                            "markdown"
                        ]
                    }
                },
                "workspace": {
                    "configuration": true,
                    "workspace_folders": true
                },
                "experimental": {
                    "hoverActions": true,
                    "commands": {
                        "commands": ["sourcepawn-vscode.gotoLocation"],
                    },
                }
            }))
            .unwrap(),
        )
        .unwrap();
    let text_document_position = test_bed.cursor().unwrap();
    let params = lsp_types::HoverParams {
        text_document_position_params: text_document_position,
        work_done_progress_params: Default::default(),
    };

    let mut res = test_bed
        .client()
        .send_request::<lsp::ext::HoverRequest>(params)
        .unwrap()
        .expect("Expected a hover response.");

    for action in &mut res.actions {
        for command in &mut action.commands {
            if let Some(args) = command.command.arguments.as_mut() {
                for arg in args.iter_mut() {
                    if let Some(arg) = arg.as_object_mut() {
                        if let Some(value) = arg.get_mut("uri") {
                            let mut uri = Url::from_str(value.as_str().unwrap()).unwrap();
                            test_bed.anonymize_uri(&mut uri);
                            *value = serde_json::Value::String(uri.to_string());
                        }
                    }
                }
            }
        }
    }

    res
}
