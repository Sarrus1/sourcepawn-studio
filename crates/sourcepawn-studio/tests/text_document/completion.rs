use insta::assert_json_snapshot;

use sourcepawn_studio::fixture::complete;

#[test]
fn global_variable_1() {
    assert_json_snapshot!(complete(
        r#"
%! main.sp
int foo;

|
^"#,
        None
    ));
}

#[test]
fn global_variable_include_1() {
    assert_json_snapshot!(complete(
        r#"
%! include/bar.sp
int foo;
%! main.sp
#include "bar.sp"

|
^"#,
        None
    ));
}

#[test]
fn global_variable_circular_include_1() {
    assert_json_snapshot!(complete(
        r#"
%! foo.sp
#include "bar.sp"
int foo;
%! bar.sp
#include "foo.sp"
int bar;

|
^"#,
        None
    ));
}

#[test]
fn include_1() {
    assert_json_snapshot!(complete(
        r#"
%! bar.sp
int bar;
%! include/baz.inc
int baz;
%! foo.sp
#include ""
          |
          ^"#,
        Some('"'.to_string())
    ));
}

#[test]
fn include_2() {
    assert_json_snapshot!(complete(
        r#"
%! bar.sp
int bar;
%! include/baz.inc
int baz;
%! foo.sp
#include <>
          |
          ^"#,
        Some('<'.to_string())
    ));
}

#[test]
fn include_3() {
    assert_json_snapshot!(complete(
        r#"
%! bar.sp
int bar;
%! include/baz.inc
#include <>
          |
          ^
%! foo.sp
int foo;"#,
        Some('<'.to_string())
    ));
}

#[test]
fn include_4() {
    assert_json_snapshot!(complete(
        r#"
%! bar.sp
int bar;
%! include/baz.inc
#include ""
          |
          ^
%! foo.sp
int foo;"#,
        Some('"'.to_string())
    ));
}

#[test]
fn include_5() {
    assert_json_snapshot!(complete(
        r#"
%! bar.sp
int bar;
%! include/baz.inc
#include ""
          |
          ^
%! include/foo.inc
int foo;"#,
        Some('"'.to_string())
    ));
}

#[test]
fn include_6() {
    assert_json_snapshot!(complete(
        r#"
%! main.sp
#include "sub_folder/foo.sp"
int main;
%! sub_folder/foo.sp
#include <sub_folder/>
                     |
                     ^
%! sub_folder/bar.sp
int bar;"#,
        Some("/".to_string())
    ));
}
