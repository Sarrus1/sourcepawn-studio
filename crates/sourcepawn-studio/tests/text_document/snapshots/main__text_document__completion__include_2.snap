---
source: crates/sourcepawn-studio/tests/text_document/completion.rs
assertion_line: 68
expression: "complete(r#\"\n%! bar.sp\nint bar;\n%! include/baz.inc\nint baz;\n%! foo.sp\n#include <>\n          |\n          ^\"#,\n    Some('<'.to_string()))"
---
[
  {
    "label": "admin",
    "kind": 17,
    "detail": "include/admin.inc",
    "insertText": "admin>"
  },
  {
    "label": "adminmenu",
    "kind": 17,
    "detail": "include/adminmenu.inc",
    "insertText": "adminmenu>"
  },
  {
    "label": "adt",
    "kind": 17,
    "detail": "include/adt.inc",
    "insertText": "adt>"
  },
  {
    "label": "adt_array",
    "kind": 17,
    "detail": "include/adt_array.inc",
    "insertText": "adt_array>"
  },
  {
    "label": "adt_stack",
    "kind": 17,
    "detail": "include/adt_stack.inc",
    "insertText": "adt_stack>"
  },
  {
    "label": "adt_trie",
    "kind": 17,
    "detail": "include/adt_trie.inc",
    "insertText": "adt_trie>"
  },
  {
    "label": "banning",
    "kind": 17,
    "detail": "include/banning.inc",
    "insertText": "banning>"
  },
  {
    "label": "bar.sp",
    "kind": 17,
    "detail": "bar.sp",
    "insertText": "bar.sp>"
  },
  {
    "label": "basecomm",
    "kind": 17,
    "detail": "include/basecomm.inc",
    "insertText": "basecomm>"
  },
  {
    "label": "baz",
    "kind": 17,
    "detail": "include/baz.inc",
    "insertText": "baz>"
  },
  {
    "label": "bitbuffer",
    "kind": 17,
    "detail": "include/bitbuffer.inc",
    "insertText": "bitbuffer>"
  },
  {
    "label": "clientprefs",
    "kind": 17,
    "detail": "include/clientprefs.inc",
    "insertText": "clientprefs>"
  },
  {
    "label": "clients",
    "kind": 17,
    "detail": "include/clients.inc",
    "insertText": "clients>"
  },
  {
    "label": "commandfilters",
    "kind": 17,
    "detail": "include/commandfilters.inc",
    "insertText": "commandfilters>"
  },
  {
    "label": "commandline",
    "kind": 17,
    "detail": "include/commandline.inc",
    "insertText": "commandline>"
  },
  {
    "label": "console",
    "kind": 17,
    "detail": "include/console.inc",
    "insertText": "console>"
  },
  {
    "label": "convars",
    "kind": 17,
    "detail": "include/convars.inc",
    "insertText": "convars>"
  },
  {
    "label": "core",
    "kind": 17,
    "detail": "include/core.inc",
    "insertText": "core>"
  },
  {
    "label": "cstrike",
    "kind": 17,
    "detail": "include/cstrike.inc",
    "insertText": "cstrike>"
  },
  {
    "label": "datapack",
    "kind": 17,
    "detail": "include/datapack.inc",
    "insertText": "datapack>"
  },
  {
    "label": "dbi",
    "kind": 17,
    "detail": "include/dbi.inc",
    "insertText": "dbi>"
  },
  {
    "label": "dhooks",
    "kind": 17,
    "detail": "include/dhooks.inc",
    "insertText": "dhooks>"
  },
  {
    "label": "entity",
    "kind": 17,
    "detail": "include/entity.inc",
    "insertText": "entity>"
  },
  {
    "label": "entity_prop_stocks",
    "kind": 17,
    "detail": "include/entity_prop_stocks.inc",
    "insertText": "entity_prop_stocks>"
  },
  {
    "label": "events",
    "kind": 17,
    "detail": "include/events.inc",
    "insertText": "events>"
  },
  {
    "label": "files",
    "kind": 17,
    "detail": "include/files.inc",
    "insertText": "files>"
  },
  {
    "label": "float",
    "kind": 17,
    "detail": "include/float.inc",
    "insertText": "float>"
  },
  {
    "label": "foo.sp",
    "kind": 17,
    "detail": "foo.sp",
    "insertText": "foo.sp>"
  },
  {
    "label": "functions",
    "kind": 17,
    "detail": "include/functions.inc",
    "insertText": "functions>"
  },
  {
    "label": "geoip",
    "kind": 17,
    "detail": "include/geoip.inc",
    "insertText": "geoip>"
  },
  {
    "label": "halflife",
    "kind": 17,
    "detail": "include/halflife.inc",
    "insertText": "halflife>"
  },
  {
    "label": "handles",
    "kind": 17,
    "detail": "include/handles.inc",
    "insertText": "handles>"
  },
  {
    "label": "helpers",
    "kind": 17,
    "detail": "include/helpers.inc",
    "insertText": "helpers>"
  },
  {
    "label": "include",
    "kind": 19,
    "detail": "include"
  },
  {
    "label": "keyvalues",
    "kind": 17,
    "detail": "include/keyvalues.inc",
    "insertText": "keyvalues>"
  },
  {
    "label": "lang",
    "kind": 17,
    "detail": "include/lang.inc",
    "insertText": "lang>"
  },
  {
    "label": "logging",
    "kind": 17,
    "detail": "include/logging.inc",
    "insertText": "logging>"
  },
  {
    "label": "mapchooser",
    "kind": 17,
    "detail": "include/mapchooser.inc",
    "insertText": "mapchooser>"
  },
  {
    "label": "menus",
    "kind": 17,
    "detail": "include/menus.inc",
    "insertText": "menus>"
  },
  {
    "label": "nextmap",
    "kind": 17,
    "detail": "include/nextmap.inc",
    "insertText": "nextmap>"
  },
  {
    "label": "profiler",
    "kind": 17,
    "detail": "include/profiler.inc",
    "insertText": "profiler>"
  },
  {
    "label": "protobuf",
    "kind": 17,
    "detail": "include/protobuf.inc",
    "insertText": "protobuf>"
  },
  {
    "label": "regex",
    "kind": 17,
    "detail": "include/regex.inc",
    "insertText": "regex>"
  },
  {
    "label": "sdkhooks",
    "kind": 17,
    "detail": "include/sdkhooks.inc",
    "insertText": "sdkhooks>"
  },
  {
    "label": "sdktools",
    "kind": 17,
    "detail": "include/sdktools.inc",
    "insertText": "sdktools>"
  },
  {
    "label": "sdktools_client",
    "kind": 17,
    "detail": "include/sdktools_client.inc",
    "insertText": "sdktools_client>"
  },
  {
    "label": "sdktools_engine",
    "kind": 17,
    "detail": "include/sdktools_engine.inc",
    "insertText": "sdktools_engine>"
  },
  {
    "label": "sdktools_entinput",
    "kind": 17,
    "detail": "include/sdktools_entinput.inc",
    "insertText": "sdktools_entinput>"
  },
  {
    "label": "sdktools_entoutput",
    "kind": 17,
    "detail": "include/sdktools_entoutput.inc",
    "insertText": "sdktools_entoutput>"
  },
  {
    "label": "sdktools_functions",
    "kind": 17,
    "detail": "include/sdktools_functions.inc",
    "insertText": "sdktools_functions>"
  },
  {
    "label": "sdktools_gamerules",
    "kind": 17,
    "detail": "include/sdktools_gamerules.inc",
    "insertText": "sdktools_gamerules>"
  },
  {
    "label": "sdktools_hooks",
    "kind": 17,
    "detail": "include/sdktools_hooks.inc",
    "insertText": "sdktools_hooks>"
  },
  {
    "label": "sdktools_sound",
    "kind": 17,
    "detail": "include/sdktools_sound.inc",
    "insertText": "sdktools_sound>"
  },
  {
    "label": "sdktools_stocks",
    "kind": 17,
    "detail": "include/sdktools_stocks.inc",
    "insertText": "sdktools_stocks>"
  },
  {
    "label": "sdktools_stringtables",
    "kind": 17,
    "detail": "include/sdktools_stringtables.inc",
    "insertText": "sdktools_stringtables>"
  },
  {
    "label": "sdktools_tempents",
    "kind": 17,
    "detail": "include/sdktools_tempents.inc",
    "insertText": "sdktools_tempents>"
  },
  {
    "label": "sdktools_tempents_stocks",
    "kind": 17,
    "detail": "include/sdktools_tempents_stocks.inc",
    "insertText": "sdktools_tempents_stocks>"
  },
  {
    "label": "sdktools_trace",
    "kind": 17,
    "detail": "include/sdktools_trace.inc",
    "insertText": "sdktools_trace>"
  },
  {
    "label": "sdktools_variant_t",
    "kind": 17,
    "detail": "include/sdktools_variant_t.inc",
    "insertText": "sdktools_variant_t>"
  },
  {
    "label": "sdktools_voice",
    "kind": 17,
    "detail": "include/sdktools_voice.inc",
    "insertText": "sdktools_voice>"
  },
  {
    "label": "sorting",
    "kind": 17,
    "detail": "include/sorting.inc",
    "insertText": "sorting>"
  },
  {
    "label": "sourcemod",
    "kind": 17,
    "detail": "include/sourcemod.inc",
    "insertText": "sourcemod>"
  },
  {
    "label": "string",
    "kind": 17,
    "detail": "include/string.inc",
    "insertText": "string>"
  },
  {
    "label": "testing",
    "kind": 17,
    "detail": "include/testing.inc",
    "insertText": "testing>"
  },
  {
    "label": "textparse",
    "kind": 17,
    "detail": "include/textparse.inc",
    "insertText": "textparse>"
  },
  {
    "label": "tf2",
    "kind": 17,
    "detail": "include/tf2.inc",
    "insertText": "tf2>"
  },
  {
    "label": "tf2_stocks",
    "kind": 17,
    "detail": "include/tf2_stocks.inc",
    "insertText": "tf2_stocks>"
  },
  {
    "label": "timers",
    "kind": 17,
    "detail": "include/timers.inc",
    "insertText": "timers>"
  },
  {
    "label": "topmenus",
    "kind": 17,
    "detail": "include/topmenus.inc",
    "insertText": "topmenus>"
  },
  {
    "label": "usermessages",
    "kind": 17,
    "detail": "include/usermessages.inc",
    "insertText": "usermessages>"
  },
  {
    "label": "vector",
    "kind": 17,
    "detail": "include/vector.inc",
    "insertText": "vector>"
  },
  {
    "label": "version",
    "kind": 17,
    "detail": "include/version.inc",
    "insertText": "version>"
  },
  {
    "label": "version_auto",
    "kind": 17,
    "detail": "include/version_auto.inc",
    "insertText": "version_auto>"
  }
]
